Add the support of the HRANDFIELD command

zevin02 opened the pull request:
Close :#1510

zevin02 committed:
Add the support of the HRANDFIELD command

- Add parse for hrandfield
- Implemented the case where the requested data is a positive number
  greater than the size of the hash.

Signed-off-by: Zewen Xu <zevin9427@gmail.com>

PragmaTwice commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,52 @@ class CommandHScan : public CommandSubkeyScanBase {
+        count_ = -*parse_result;
+        uniq_ = false;
+      }
+      //如果有withvalue这个参数，就一定有前面的count
```
Please comment in English :)

Yangsx-1 commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,52 @@ class CommandHScan : public CommandSubkeyScanBase {
+    std::vector<FieldValue> field_values;
+    auto s = hash_db.RandField(args_[1], &field_values, count_, uniq_,
+                               withvalues_ ? HashFetchType::kAll : HashFetchType::kOnlyKey);
+    if (!s.ok()) {
```
It should return `nil` if the key is not exist.

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -386,5 +386,46 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  }
+  /* case3:The number of elements inside the hash is not greater than the number of
+    elements inside the hash*/
+  else {
```
is this still wip?

zevin02 replied:
I have just completed one case, and I am still working on implementing the other two cases.

zevin02 committed:
- implement hrandfield
- add test for hrandfield
- Fix output (nil) when the key does not exist

Signed-off-by: Zewen Xu <zevin9427@gmail.com>

Yangsx-1 commented on src/types/redis_hash.cc:
```diff
@@ -386,5 +387,78 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  if (!uniq) {
+    field_values->reserve(count);
+    for (uint64_t i = 0; i < count; i++) {
+      uint64_t index = (std::rand() % size);
```
It's better to add a rand seed before `rand()` and use `std::mt19937` the same as below.

mapleFU replied:
Seems that we have a `src/common/rand.h`, could we use that @PragmaTwice

Yangsx-1 commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,52 @@ class CommandHScan : public CommandSubkeyScanBase {
+ private:
+  bool withvalues_ = false;
+  bool uniq_ = true;
+  uint64_t count_ = 1;
```
Maybe `int64_t` is better.

Yangsx-1 commented on src/types/redis_hash.cc:
```diff
@@ -25,11 +25,12 @@
 #include "db_util.h"
 #include "parse_util.h"
-
+#include "time_util.h"
```
Leave a blank line here, the same as the end of namespace and between functions.   :)

Yangsx-1 commented on src/types/redis_hash.cc:
```diff
@@ -386,5 +387,78 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+    if (type == HashFetchType::kAll) {
+      values.emplace_back(iter->value().ToString());
+    }
+  }
```
You can use `GetAll` function, it's almost the same.

Yangsx-1 commented:
You can also add some go tests.

Yangsx-1 commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,52 @@ class CommandHScan : public CommandSubkeyScanBase {
+
+    auto s = hash_db.RandField(args_[1], &field_values, count_, uniq_,
+                               withvalues_ ? HashFetchType::kAll : HashFetchType::kOnlyKey);
+    if (!s.ok() && !s.IsNotFound()) {
```
Sorry, in redis, if without the additional count argument, the command returns a Bulk Reply with the randomly selected field, or nil when key does not exist. When the additional count argument is passed, the command returns an array of fields, or an empty array when key does not exist. If the WITHVALUES modifier is used, the reply is a list fields and their values from the hash. See details in https://redis.io/commands/hrandfield/

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -386,5 +387,78 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  read_options.iterate_upper_bound = &upper_bound;
+  storage_->SetReadOptions(read_options);
+
+  uint64_t size = metadata.size;
```
Getting all values in Hash might be heavy, could we lazy load these values later, or just leave it here and add a TODO for further optimizations?

zevin02 replied:
I can add a TODO for further optimizations

git-hulk replied:
Why not use `GetAll` to get all fields for field_values directly?

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -386,5 +387,78 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+      values.emplace_back(iter->value().ToString());
+    }
+  }
+
```
Should we optimize the case when `count == 0`, when `count == 0`, seems that no values need to be loaded from storage

Yangsx-1 replied:
Yes, it's a good optimization.  :)

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -386,5 +387,78 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  else if (size <= count) {
+    field_values->reserve(size);
+    for (uint64_t i = 0; i < size; i++) {
+      if (type == HashFetchType::kAll) {
```
```c++
      if (type == HashFetchType::kAll) {
        field_values->emplace_back(fields[i], values[i]);
      } else {
        field_values->emplace_back(fields[i], "");
      }
```

This part occurs three times here, would you mind encapsulate it to a lambda?

mapleFU reviewed (commented):
I didn't review the `Hash::RandField` implements, just some style nits

PragmaTwice commented on src/types/redis_hash.cc:
```diff
@@ -386,5 +387,78 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+
+  // Case 2: Requested count is greater than or equal to the number of elements inside the hash
+  else if (size <= count) {
+    field_values->reserve(size);
```
no need to shuffle here? what's redis corresponding behavior?

zevin02 replied:
In the redis implementation, when the requested data is larger than the data stored in the hash, the data in the hash is returned in order

PragmaTwice replied:
oh got it

PragmaTwice commented on src/types/redis_hash.cc:
```diff
@@ -386,5 +387,78 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  }
+  // Case 3: Requested count is less than the number of elements inside the hash
+  else {
+    field_values->reserve(count);
+    std::vector<uint64_t> indices(size);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::shuffle(indices.begin(), indices.end(), gen);  // use Fisher-Yates shuffle algorithm to randomize the order
+    for (uint64_t i = 0; i < count; i++) {
+      uint64_t index = indices[i];
+      if (type == HashFetchType::kAll) {
+        field_values->emplace_back(fields[index], values[index]);
+      } else {
+        field_values->emplace_back(fields[index], "");
+      }
+    }
+  }
```
seems the output will always be a random order of index {0,...`count`-1}, which cannot be called "rand field" IMHO.

PragmaTwice replied:
if you want to do the shuffle, you need to shuffle [0, `hash_elem_size`) and cut the first `count` numbers as indices.

zevin02 replied:
> if you want to do the shuffle, you need to shuffle [0, `hash_elem_size`) and cut the first `count` numbers as indices.

should I modify it like this, or what should I do
~~~c
std::shuffle(indices.begin(), indices.begin() + count, gen);  
std::sort(indices.begin(), indices.begin() + count);
~~~

zevin02 replied:
> if you want to do the shuffle, you need to shuffle [0, `hash_elem_size`) and cut the first `count` numbers as indices.
should I modify like this,or anything else
~~~cpp
indices.resize(count);
~~~

zevin02 replied:
> if you want to do the shuffle, you need to shuffle [0, `hash_elem_size`) and cut the first `count` numbers as indices.

should I modify like this,or anything else
~~~cpp
indices.resize(count);
~~~

PragmaTwice replied:
oh I may mistake `std::vector indices(size)` as `indices(count)`, sorry for that. the implementation looks fine to me.

PragmaTwice replied:
I think you can just use random_device as the URBG without mt19937. e.g 

```
shuffle(begin, end, std::random_device{})
```

Yangsx-1 replied:
A little reference for [why not just use random_device](https://stackoverflow.com/questions/39288595/why-not-just-use-random-device).  :)

mapleFU replied:
usally random device can be used to generate a mt19937 seed

PragmaTwice replied:
> A little reference for [why not just use random_device](https://stackoverflow.com/questions/39288595/why-not-just-use-random-device). :)

@maplefu @Yangsx-1  I am not sure whether the common way to implement std::random_device is still `/dev/urandom`, the answer seems to be out-dated.

Usually, e.g. in x86, the `rdrand` instruction is a far better way to implement std::random_device with great performance.

PragmaTwice replied:
> usally random device can be used to generate a mt19937 seed

mt19937 is just a good PRNG algorithm. it is not required to use if your device can generate **real random numbers**.

PragmaTwice replied:
@Yangsx-1 I have checked the latest version of libstdc++, which are using `rdrand` when possible.

https://github.com/gcc-mirror/gcc/blob/63ae6bc60c0f67fb2791991bf4b6e7e0a907d420/libstdc%2B%2B-v3/src/c%2B%2B11/random.cc#L34

mapleFU replied:
```
/* If our unsigned long type can store a 64 bit number, use a 64 bit PRNG. */
#if ULONG_MAX >= 0xffffffffffffffff
#define randomULong() ((unsigned long) genrand64_int64())
#else
#define randomULong() random()
#endif
```

@PragmaTwice after check in redis, I guess it would uses `randomULong()`, and it wraps a PRNG algorithm. I guess mt19937 would be ok

zevin02 commented:
> You can also add some go tests.

I will do it later

zevin02 committed:
- fix reply format
- call the GetAll function to get all elements
- add some cpp tests
- wrap functions with lambdas
- updated random selection

Signed-off-by: Zewen Xu <zevin9427@gmail.com>

mapleFU commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,54 @@ class CommandHScan : public CommandSubkeyScanBase {
+      }
+      l_ = *parse_result;
+
+      if (args.size() > 4 || (args.size() == 4 && strcasecmp(args[3].c_str(), "withvalues"))) {
```
```suggestion
      if (args.size() > 4 || (args.size() == 4 && !util::EqualICase(args[3], "withvalues"))) {
```

mapleFU commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,54 @@ class CommandHScan : public CommandSubkeyScanBase {
+
+ private:
+  bool withvalues_ = false;
+  int64_t l_ = 0;
```
`l_` is a confusing name, I guess `count_` could be much better

zevin02 replied:
`count` is also used in the `RandField` , and `l` is also used in parameter extraction in the implementation of redis, so I also use `l`. Is there any better suggestion?

mapleFU replied:
Hmm I guess Redis Command level uses name `COUNT`. Maybe using `command_count` or other is ok?

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,60 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+
+  std::string ns_key;
+  AppendNamespacePrefix(user_key, &ns_key);
+  HashMetadata metadata(false);
```
```suggestion
  HashMetadata metadata(/*generate_version=*/false);
```

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,60 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  uint64_t size = metadata.size;
+  std::vector<FieldValue> samples;
+  // TODO: Getting all values in Hash might be heavy, consider lazy-loading these values later
+
```
Can just return when `l == 0`?

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,60 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+
+  GetAll(user_key, &samples, type);
+
+  auto processField = [field_values, &samples, type](uint64_t index) {
```
would `appendFieldWithIndex` or others be better?

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,60 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+      field_values->emplace_back(samples[index].field, samples[index].value);
+    } else {
+      field_values->emplace_back(samples[index].field, "");
+    }
```
By the way, could we reserve the `field_values` as `min(size, count)` in this case after `GetAll` called?

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,60 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+    }
+  };
+
+  if (!unique) {
```
would it ok to use

```
if (!unique || count == 1)
```

like in redis? I believe it also handles the `noparameter` case

mapleFU reviewed (commented):
Looks much better now!

zevin02 committed:
update

Signed-off-by: Zewen Xu <zevin9427@gmail.com>

zevin02 committed:
add go test for hrandfield

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,58 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  uint64_t size = metadata.size;
+  std::vector<FieldValue> samples;
+  // TODO: Getting all values in Hash might be heavy, consider lazy-loading these values later
+  if (count == 0 && !noparmeter) return rocksdb::Status::OK();
```
Would we remove `noparameter` in `Hash::RandField` by default set `command_count_ == 1`?

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,58 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  } else {
+    // Case 3: Requested count is less than the number of elements inside the hash
+    std::vector<uint64_t> indices(size);
+    std::iota(indices.begin(), indices.end(), 0);
```
Seems C++17 provides a `sample` ( https://en.cppreference.com/w/cpp/algorithm/sample ) , would it be better ? @PragmaTwice

mapleFU commented on src/types/redis_hash.h:
```diff
@@ -61,6 +61,8 @@ class Hash : public SubKeyScanner {
   rocksdb::Status Scan(const Slice &user_key, const std::string &cursor, uint64_t limit,
                        const std::string &field_prefix, std::vector<std::string> *fields,
                        std::vector<std::string> *values = nullptr);
+  rocksdb::Status RandField(const Slice &user_key, std::vector<FieldValue> *field_values, int64_t l, bool noparmeter,
```
```suggestion
  rocksdb::Status RandField(const Slice &user_key, int64_t command_count, bool noparmeter, std::vector<FieldValue> *field_values, 
```

Seems that input field should better before output fields

mapleFU commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,58 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+  std::vector<FieldValue> samples;
+  // TODO: Getting all values in Hash might be heavy, consider lazy-loading these values later
+  if (count == 0 && !noparmeter) return rocksdb::Status::OK();
+  GetAll(user_key, &samples, type);
```
`GetAll` returns a `Status`, seems that we'd better check it.

mapleFU reviewed (commented):
Looks much better! I didn't check the tests carefully

zevin02 committed:
update

zevin02 committed:
fix format

PragmaTwice commented:
Seems here are some clang-tidy reports that need to be fixed.

https://github.com/apache/kvrocks/actions/runs/5542464594/jobs/10117205769?pr=1565

zevin02 committed:
fix format

mapleFU commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,54 @@ class CommandHScan : public CommandSubkeyScanBase {
+    std::vector<std::string> result_entries;
+    result_entries.reserve(field_values.size());
+    for (const auto &p : field_values) {
+      result_entries.emplace_back(p.field);
```
would `std::move(p.field)` be better here?

mapleFU reviewed (commented):
Rest LGTM!

By the way, I noticed that you have written:

```
// TODO: Add test to verify randomness of the selected random fields
```

Would this be added in this patch or later?

zevin02 commented:
> 

This patch will not implement that

zevin02 commented:
> Rest LGTM!
> 
> By the way, I noticed that you have written:
> 
> ```
> // TODO: Add test to verify randomness of the selected random fields
> ```
> 
> Would this be added in this patch or later?

This patch will not implement that

mapleFU commented on tests/gocase/unit/type/hash/hash_test.go:
```diff
@@ -783,6 +783,53 @@ func TestHash(t *testing.T) {
+			require.Equal(t, []string{"key1", "key2", "key3"}, result1)
+			result3, err3 := rdb.HRandField(ctx, testKey, 2).Result()
+			require.NoError(t, err3)
+			require.Len(t, result3, 2)
```
would you mind check the value is in `{"key1", "value1", "key2", "value2", "key3", "value3"}`?

mapleFU commented on tests/gocase/unit/type/hash/hash_test.go:
```diff
@@ -783,6 +783,53 @@ func TestHash(t *testing.T) {
+			testKey := "test-hash-1"
+			require.NoError(t, rdb.Del(ctx, testKey).Err())
+			require.NoError(t, rdb.HSet(ctx, testKey, "key1", "value1", "key2", "value2", "key3", "value3").Err())
+			result1, err1 := rdb.HRandField(ctx, testKey, 5).Result()
```
can you use same err and result to avoid too many variable names?

zevin02 replied:
Ok, is there anything else that needs to be improved?

mapleFU replied:
I guess no, if these are finished I'll approve this pull request. Thanks for your working

zevin02 committed:
update

mapleFU reviewed (approved):
cc @PragmaTwice @Yangsx-1 @git-hulk

PragmaTwice commented on tests/cppunit/types/hash_test.cc:
```diff
@@ -331,3 +331,34 @@ TEST_F(RedisHashTest, HRangeByLexNonExistingKey) {
+  EXPECT_TRUE(s.ok() && fvs.size() == 0);
+
+  hash_->Del(key_);
+}
```
```suggestion
}

```

zevin02 committed:
update

Yangsx-1 commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,54 @@ class CommandHScan : public CommandSubkeyScanBase {
     return Status::OK();
   }
 };
+class CommandHRandField : public Commander {
```
```suggestion

class CommandHRandField : public Commander {
```

Yangsx-1 commented on src/commands/cmd_hash.cc:
```diff
@@ -376,6 +376,54 @@ class CommandHScan : public CommandSubkeyScanBase {
+      }
+    }
+    return Commander::Parse(args);
+  }
```
```suggestion
  }
  
```

Yangsx-1 commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,57 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+    std::iota(indices.begin(), indices.end(), 0);
+    std::shuffle(indices.begin(), indices.end(),
+                 std::random_device{});  // use Fisher-Yates shuffle algorithm to randomize the order
+    indices.resize(count);
```
This resize may be not necessary.

Yangsx-1 commented on src/types/redis_hash.cc:
```diff
@@ -387,4 +388,57 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
+    }
+  }
+  return rocksdb::Status::OK();
+}
```
```suggestion
}

```

Yangsx-1 reviewed (commented):
Others are good to me.  :)

zevin02 committed:
update

Yangsx-1 commented:
> fixes :#1510
> 
> * Add parse for hrandfield
> * Implemented the case where the requested data is a positive number greater than the size of the hash.

Close #1510 is better.

zevin02 committed:
update

git-hulk committed:
Merge branch 'unstable' into hrandfield-command

git-hulk commented:
@zevin02 Thanks for your contribution and patience and greater thanks to reviewers: @mapleFU @PragmaTwice @Yangsx-1
